mirgecom/mpi.py

```diff
@@ -32,6 +32,64 @@
 import sys
 
 
+def _check_gpu_oversubscription():
+    """
+    Check whether multiple ranks are running on the same GPU on each node.
+
+    Only works with CUDA devices currently due to the use of the
+    PCI_DOMAIN_ID_NV extension.
+    """
+    from mpi4py import MPI
+    import pyopencl as cl
+
+    size = MPI.COMM_WORLD.Get_size()
+    rank = MPI.COMM_WORLD.Get_rank()
+
+    if size <= 1:
+        return
+
+    cl_ctx = cl.create_some_context()
+    dev = cl_ctx.devices
+
+    # No support for multi-device contexts
+    if len(dev) > 1:
+        raise RuntimeError(f"Multiple devices selected for rank {rank}.")
+
+    dev = dev[0]
+
+    if not (dev.type & cl.device_type.GPU):
+        return
+
+    node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+    node_rank = node_comm.Get_rank()
+
+    try:
+        domain_id = hex(dev.pci_domain_id_nv)
+    except (cl._cl.LogicError, AttributeError):
+        from warnings import warn
+        warn("Cannot detect whether multiple ranks are running on the"
+             " same GPU because it requires Nvidia GPUs running with"
+             " pyopencl>2021.1.1 and (Nvidia CL or pocl>1.6).")
```

Comment:
It will say this for AMD and Intel GPUs now. Is that intended?

Reply:
Do you mean this warning should not be shown for AMD/Intel GPUs? Or is the text unclear?

Reply:
Fair enough.